L1Trigger/TrackFindingTracklet/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,261 @@
+    setup_ = &iSetup.getData(esGetTokenSetup_);
+    if (!setup_->configurationSupported())
+      return;
+    // check process history if desired
+    if (iConfig_.getParameter<bool>("CheckHistory"))
+      setup_->checkHistory(iRun.processHistory());
+    // helper class to extract structured data from TTDTC::Frames
+    dataFormats_ = &iSetup.getData(esGetTokenDataFormats_);
+    // helper class to encode layer
+    layerEncoding_ = &iSetup.getData(esGetTokenLayerEncoding_);
+    // helper class to assign tracks to channel
+    trackBuilderChannel_ = const_cast<TrackBuilderChannel*>(&iSetup.getData(esGetTokenTrackBuilderChannel_));
+  }
+
+  void ProducerKFin::produce(Event& iEvent, const EventSetup& iSetup) {
+    auto toFrameStub = [](StubKFin* stub) { return stub->frame(); };
+    auto toFrameTrack = [](const TrackKFin& track){ return track.frame(); };
+    const DataFormat& dfCot = dataFormats_->format(Variable::cot, Process::sf);
+    const DataFormat& dfZT = dataFormats_->format(Variable::zT, Process::sf);
+    const DataFormat& dfQoverPt = dataFormats_->format(Variable::qOverPt, Process::sf);
+    const DataFormat& dfPhiT = dataFormats_->format(Variable::phiT, Process::sf);
+    const DataFormat& dfPhi = dataFormats_->format(Variable::phi, Process::sf);
+    const DataFormat& dfZ = dataFormats_->format(Variable::z, Process::sf);
+    const int numStreamsTracks = setup_->numRegions() * trackBuilderChannel_->numChannels();
+    const int numStreamsStubs = numStreamsTracks * setup_->numLayers();
+    // empty KFin products
+    TTDTC::Streams streamAcceptedStubs(numStreamsStubs);
+    StreamsTrack streamAcceptedTracks(numStreamsTracks);
+    TTDTC::Streams streamLostStubs(numStreamsStubs);
+    StreamsTrack streamLostTracks(numStreamsTracks);
+    // read in hybrid track finding product and produce KFin product
+    if (setup_->configurationSupported()) {
+      Handle<TTTracks> handleTTTracks;
+      iEvent.getByToken<TTTracks>(edGetTokenTTTracks_, handleTTTracks);
+      const TTTracks& ttTracks = *handleTTTracks.product();
+      vector<TTTrackRefs> ttTrackRefsStreams(numStreamsTracks);
+      vector<int> nTTTracksStreams(numStreamsTracks, 0);
+      int channelId;
+      for (const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack : ttTracks)
+        if (trackBuilderChannel_->channelId(ttTrack, channelId))
+          nTTTracksStreams[channelId]++;
+      channelId = 0;
+      for (int nTTTracksStream : nTTTracksStreams)
+        ttTrackRefsStreams[channelId++].reserve(nTTTracksStream);
+      int i(0);
+      for (const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack : ttTracks)
+        if (trackBuilderChannel_->channelId(ttTrack, channelId))
+          ttTrackRefsStreams[channelId].emplace_back(TTTrackRef(handleTTTracks, i++));
+      for (channelId = 0; channelId < numStreamsTracks; channelId++) {
+        const TTTrackRefs& ttTrackRefs = ttTrackRefsStreams[channelId];
+        vector<TrackKFin> tracks;
```

Comment:
Add comment: "Create vector of stubs/tracks in KF format from TTTracks".

Reply:
done

Reply:
Has this comment been added?

Reply:
yes L163